docs/design-proposals/2022-04 Parquet.md

```diff
@@ -0,0 +1,256 @@
+
+```
+repeated group Attrs {
+    required binary Key (STRING);
+    required binary Value (STRING);
+}
+```
+
+### Compression and Encoding
+Parquet has robust support for many compression algorithms and data encodings. We have found excellent combinations of storage size and performance with the following:
+
+1. LZ4 Compression - Enable on all columns
+2. Dictionary encoding - Enable on all string columns (including byte array ParentSpanID). Most strings are very repetitive so this works well to optimize storage size.  However we can greatly speed up search by inspecting the dictionary first and eliminating pages with no matches.
+3. Time and duration unix nanos - Delta encoding
+4. Rarely used columns such as DroppedAttributesCount - These columns are usually all zeroes, RLE works well.
+
+### Bloom Filters
+Parquet has native support for bloom filters however we are not using them at this time.  Tempo already has sophisticated support for sharding and caching bloom filters, and we will continue to leverage that for now.
+
+## Results from Local Testing
+Here are some interesting column sizes from a block of size ~600MB containing ~150K traces:
+
+```
+column.Trace.DurationNanos                                                    154414 values  0.64 MB
+column.Trace.ResourceSpans.Resource.ServiceName                              2672245 values  2.01 MB
+column.Trace.ResourceSpans.InstrumentationLibrarySpans.Spans.HttpStatusCode 10428415 values  6.35 MB 
+column.Trace.ResourceSpans.InstrumentationLibrarySpans.Spans.ID             10428415 values 82.37 MB  # this is never used!
+```
+
+Some super early benchmarks on local SSDs:
+
+* Searching protobuf data for a simple query `cluster=ops and minDuration=1s`
+
+```
+=== RUN   TestSearchProto
+Traces : 55
+Traces inspected: 154414
+--- PASS: TestSearchProto (21.11s)
+```
+
+* Same query on Parquet blocks
+
+```
+=== RUN   TestSearchPipelineStatic
+Traces : 55
+Traces inspected: 154414
+Reads  : 290 6.65 MB
+--- PASS: TestSearchPipelineStatic (0.18s)
+```
+
+## Implementation
+We have been refactoring Tempo recently in anticipation of this, therefore Parquet-formatted blocks should be as easy as creating a new folder `/tempodb/encoding/vparquet` and implementing the `VersionedEncoding` interface. Tempo's parquet support will be based on the new library https://github.com/segmentio/parquet-go which we have used for all prototyping and have had execellent success with.
+
+### Write path
+
+* Creating blocks: The ingester will write and flush parquet-formatted blocks. The compactor will only compact like-encoded-blocks, i.e. parquet to parquet. This seems better than having the ingester continue to flush proto and the compactor to convert them.
+
+*  One tunable is the row group size.  Our index pages target something like 256K, but row groups are typically much larger, 50-100MB.  Not sure about the best value here, will need to experiment.
```

Comment:
I'd say this difference is concerning. The smaller pages allows for heavy parallelization across the same block. My guess is that the significantly reduced i/o will more than make up for it. Agree experimentation is the best path to finding a good value.

Reply:
While we could reduce the rowgroup size further and get a lot more workers active on a single block, I agree with you that the reduced I/O makes up for the reduced parallelism. A lot of our local benchmarking uses sequential scanning of all rowgroups and the results are pretty good already. We'll continue to experiment.

Reply:
Also, the columnar format allows adding a new dimension to parallelism -- which is to search on multiple columns simultaneously.

Reply:
Good thoughts here.  Adding a note that we default to 10MB per search shard, which seems like a reasonable lower bound for row group size experimentation.  Also our default index downsampling is 1MB not 256KB, will fix.